### PR TITLE
[inductor] Unroll small reductions

### PR DESCRIPTION
**Pull request by jansel**

### CODE REVIEW

File: torchinductor/ir.py [hunk 1]
```diff
@@ -582,6 +657,18 @@ def fn(index):
 
             return Pointwise.create(device, dst_dtype, fn, ranges)
 
+        if (
+            V.graph.sizevars.size_hint(reduction_numel)
```
Comment (Chillee) [inline torchinductor/ir.py:661]: I wonder if this is a good idea. If the reduction dim suddenly changes with dynamic shapes, it'll probably be quite bad if we unroll a massive reduction.

Perhaps we should only do this if the reduction dim is static?

Comment (jansel) [reply torchinductor/ir.py:661]: This won't happen due to the guard.  

"Is adding guards ok for optimizations?" is a reasonable question though.  We do it in some other places too.

File: torchinductor/ir.py [hunk 2]
```diff
@@ -562,6 +562,81 @@ def outer_reduction_splits(reduction_numel_hint, numel_hint):
...
+                )
+
+        else:
+            raise NotImplementedError(f"unknown reduction_type={reduction_type}")
+
+        def fn(index):
+            return functools.reduce(
+                combine_fn,
+                (
+                    value_fn(index, rindex)
```
Comment (Chillee) [inline torchinductor/ir.py:613]: I find this ordering a bit confusing, since we're defining `value_fn` after this.

**Review by Chillee — APPROVED** [review]
LGTM

File: torchinductor/ir.py (hunk 2 above)
Comment (jansel) [reply torchinductor/ir.py:613]: Yeah, I had the other way before, but it results in extra control flow which also doesn't look nice.

**Pull request merged**